src/Model/LanguageFeatures.re

```diff
@@ -0,0 +1,77 @@
+/*
+ * LanguageFeatures.re
+ *
+ * This module documents & types the protocol for communicating with the VSCode Extension Host.
+ */
+
+open Oni_Core;
+
+module CompletionProvider = {
+  type t =
+    (Buffer.t, CompletionMeet.t, Position.t) =>
+    option(Lwt.t(list(CompletionItem.t)));
+};
+
+type t = {completionProviders: list(CompletionProvider.t)};
+
+let empty = {completionProviders: []};
```

Comment:
This type gets stored in the `State` - and there are some functional values here (as the `CompletionProvider.t` is a function). @glennsl - I believe you had some concerns about storing functions on the `State` (if I recall correctly), so wanted to run this by you.

Reply:
Yeah, functions don't have a "structure" like other values. They can't be printed, they can't be compared and as closures can hold state that you cannot see, which can cause bugs that are very hard to track down. If you try to compare functions structurally it will just crash, which isn't great, and might also remove opportunities for optimizations like only rendering the parts of the UI whose state has changed for example.

I also think it's largely unnecessary. Unless you need dynamic linking you'll have all the functions available at compile-time. If you need "Inversion of Control" or "dependency injection" you can do so using partial application, a closure or a functor along with some associated value, a string, generated unique id or proper variant, that can be put in the model and actions.

Or if the limitations of interacting with an imperative API makes that difficult, it might be better to treat the entire thing as external. Abstract it so you have a single completion provider that can register other providers or be swapped out with a dummy. Request completions through an effect instead of an ordinary function and deliver the results via an action.

Or isolate the imperativeness to the extension host API, if that's likely the only "culprit", by having a single completion provider for the extension host which handles registering and such.

Reply:
Thanks for the details @glennsl ! Really appreciate your thoughts.

These concerns make sense to me:
1) Can't be printed
2) Can't be compared
3) Hold invisible state / hard-to-track bugs
4) Can't compare state
5) Can't persist state
6) Breaks incremental rendering

It seems like having an 'id' could address 1 & 2, and having logging around / in could help 3.

The comparing the state / can't persist state 4 & 5 is a price that we pay - but the comparison today we only are using a reference equality (shallow equality) check at the top-level, so I'm not as concerned about that piece (persistence would need to be custom - we won't persist the whole state anyway, but maybe subsets / pieces of it).

The incremental rendering is a good point, but not something we support yet.

> If you need "Inversion of Control" or "dependency injection" you can do so using partial application, a closure or a functor along with some associated value, a string, generated unique id or proper variant, that can be put in the model and actions.

Wouldn't a closure or a functor have the same issues? Trying to understand the difference - is it having the associated value / identifier / some proper variant that can at least help with logging?

> Abstract it so you have a single completion provider that can register other providers or be swapped out with a dummy.

This is (sort of) what we have with this refactoring - the `LanguageFeatures.requestCompletions` is the 'aggregate provider' that registers other providers. The key idea is what constitutes a completion provider across various strategies? Like the extension host has one way of getting completions (call `provideSuggestions` API), and keyword completion has another (consulting a per-buffer, per-filetype map to see what keywords are available) - the simplest common way that I can see, right now, to express these is a function. Pretty easy for both those places to implement that completion provider function and register it.

> Or isolate the imperativeness to the extension host API, if that's likely the only "culprit", by having a single completion provider for the extension host which handles registering and such.

This was essentially the pre-refactoring design, if I understand correctly - but it doesn't extend well to things like keyword completion (registering a separate provider that can get completions from syntax highlighting)

Reply:
> ... the comparison today we only are using a reference equality (shallow equality) check at the top-level, so I'm not as concerned about that piece

I'd be more concerned about a rogue unintentional structural equality check causing a crash. That can very easily happen. Though it might be an effective way of revealing what might otherwise have been a performance problem. But I think users will prefer a minor performance hit over a crash.

> Wouldn't a closure or a functor have the same issues?

If you put them in the state they would. But that's not what I'm suggesting.

> is it having the associated value / identifier / some proper variant that can at least help with logging?

No, I mean as a way of doing dependency injection without putting the dependencies in the state. From what I can see there are two different dependency relations though, and they seem pretty entangled here, so let me try to untangle and explain them separately:

One is between the model/store and the "first level" of completion providers such as `ExtensionCompletionProvider` and `KeywordCompletionProvider`. You probably don't want the model and store to depend on them, and you want to be able to switch them out for testing. For this you can use a functor, partial application or closure to inject the dependencies in `StoreThread` when you wire everything up. This is done elsewhere where `*StoreConnector.start` takes dependencies as arguments and the updater closes over them.

The other dependency relation is really just between `ExtensionCompletionProvider` and the extensions themselves, which, as I understand, uses the `onSuggestionProvider` to register themselves. And it does so by providing an `id` and a set of capabilities, not a callback function. So I'm not sure why you're turning it into a function and storing that, when the id and capabilities could be stored instead.

However, if for some reason you _do_ need something like a registry of functions, what I'm suggesting is to keep the complexities of that external to the general CompletionProvider mechanism. Keep the dirt in the corner instead of spreading it around the room. It ought to be a concern solely for the specific `CompletionProvider`, which then provides a single interface for all extension completion providers. Treat it as a side-effect that just dispatches an action with results when it's done.

> the simplest common way that I can see, right now, to express these is a function

Functions may be easy, but I don't think they're simple. It's probably the most complex construct in the language, which is why they're so hard to serialize and compare, and because they can do pretty much anything they're hard to debug. Data is simple, and you can easily map data to code. If polymorphism is needed, a GADT with a constructor for each completion provider with an existential to store provider-specific state is much simpler than a function, but definitely not as easy. From what I can tell a simple variant should be sufficient though. Then you select the code to run by using a simple switch:

```rust
switch (completionProvider) {
| Extension(id) => ExtensionCompletionProvider.getCompletions(id)
| Keyword(state) => KeywordCompletionProvider.getCompletions(state)
};
```

I don't see why it should need to be any more complicated than this, though of course that certainly doesn't mean there isn't good reason.

Reply:
Thanks for the thoughtful & detailed response, @glennsl !

I think this is the crux of our different perspectives:

> Functions may be easy, but I don't think they're simple. It's probably the most complex construct in the language, which is why they're so hard to serialize and compare, and because they can do pretty much anything they're hard to debug. Data is simple, and you can easily map data to code. If polymorphism is needed, a GADT with a constructor for each completion provider with an existential to store provider-specific state is much simpler than a function, but definitely not as easy.

In the current iteration of design w/ this feature - at this time, I do think that the functions are a simpler expression of these integrations than the proposed alternatives (GADT or variants), even with the trade-offs we've discussed, and will be the easiest path forward (at least for me!) to implement the ExtHost integration. 
 
But it certainly is very helpful to review the trade-offs and examine alternate designs, and I am open to refactoring to an alternate design, especially as we have more examples in place.

So I'd prefer to move forward with the function approach for now - but I'm open to revisiting in the future - I put more details in a card specific to the 'functional' parts of this module: https://github.com/onivim/oni2/projects/10#card-30447137